File: 0213-house-robber-ii/0213-house-robber-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
  const n = nums.length;
  
  return n === 1 ? nums[0] : Math.max(helper(0, n - 1, nums), helper(1, n, nums));
};

const helper = (start, end, nums) => {
  let [prev1, prev2] = [0, 0];
  
  for (let i = start; i < end; i++) {
    [prev1, prev2] = [prev2, Math.max(prev1 + nums[i], prev2)];
  }
  
  return prev2;
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def dfs(start, end):
            a = b = 0
            for i in range(start, end):
                a, b = b, max(b, nums[i] + a)
            return max(a, b)
          
          
        n = len(nums)
        return nums[0] if n == 1 else max(dfs(0, n - 1), dfs(1, n))